Two phase deactivate

njalerikson opened the pull request:
This is the PR for #3915.

As discussed there this PR solves the `{de,}activate.d` scripts' issue of being unable to change the `$PATH` by introducing a two phase deactivate.

Two phase because upon the first call, if there are `deactivate.d` scripts then the first phase is to source those scripts and then to do a second conda call - `deactivatepost`. The second phase, the deactivate post, will then perform the final unsetting, prompt modifications, and `$PATH` modification. If however there are no `deactivate.d` scripts then there will only be a single deactivate phase.

This is a relatively dumb fix in my opinion, primarily because I've statically embedded the `deactivatepost` command in the Activator class for each shell. I would love to hear any smarter ideas.

As discussed with @kalefranz, this is not an ideal fix but rather a quick fix until other longterm ideas become reality.

mbargull commented on conda/activate.py:
```diff
@@ -99,6 +105,8 @@ def __init__(self, shell, arguments=None):
             self.export_var_tmpl = '@SET "%s=%s"'
             self.run_script_tmpl = '@CALL "%s"'
 
+            self.deactivatepost_tmpl = """@CALL %_CONDA_EXE% shell.cmd.exe deactivatepost"""
```
For `cmd.exe`, I think  we'd need the monstrosity of
https://github.com/conda/conda/blob/4.5.0rc0/conda/shell/Library/bin/conda.bat#L45
https://github.com/conda/conda/blob/4.5.0rc0/conda/shell/Library/bin/conda.bat#L50-L52
(Maybe with a slightly different variable name for `_TEMP_SCRIPT_PATH` to avoid collision?)

njalerikson replied:
wow I didn't even process how ugly the `cmd.exe` version was...

mbargull commented on conda/activate.py:
```diff
@@ -111,6 +119,8 @@ def __init__(self, shell, arguments=None):
             self.export_var_tmpl = 'set -gx %s "%s"'
             self.run_script_tmpl = 'source "%s"'
 
+            self.deactivatepost_tmpl = """eval $_CONDA_EXE shell.fish deactivatepost"""
```
For `fish` we need two `eval`s, acc. to https://github.com/conda/conda/blob/4.5.0rc0/conda/shell/etc/fish/conf.d/conda.fish#L92.

mbargull commented on conda/activate.py:
```diff
@@ -291,6 +308,28 @@ def build_activate(self, env_name_or_prefix):
 
     def build_deactivate(self):
         # query environment
+        try:
```
Better avoid a large `try: ... except` block, i.e.,
```python
try:
    old_conda_prefix = self.environ['CONDA_PREFIX']
except KeyError:
    pass
else:
    ...
```
or (more consistent with the other code in this file)
```python
old_conda_prefix = self.environ['CONDA_PREFIX']
if old_conda_prefix is not None:
    ...
```

mbargull commented on conda/activate.py:
```diff
@@ -147,6 +157,10 @@ def deactivate(self):
+        return self._finalize(self._yield_commands(self.build_deactivatepost()),
+                              self.tempfile_extension)
+
     def reactivate(self):
```
For `reactivate` we also need a two-staged process if we have `deactivate` scripts.

mbargull commented on conda/activate.py:
```diff
@@ -62,6 +62,8 @@ def __init__(self, shell, arguments=None):
             self.set_var_tmpl = "%s='%s'"
             self.run_script_tmpl = '\\. "%s"'
 
+            self.deactivatepost_tmpl = """eval "$(PS1="$PS1" $_CONDA_EXE shell.posix deactivatepost)" """
```
Use `\eval` to guard against potential `alias`es.

mbargull commented on conda/activate.py:
```diff
@@ -87,6 +91,8 @@ def __init__(self, shell, arguments=None):
             self.export_var_tmpl = "$%s = '%s'"
             self.run_script_tmpl = 'source "%s"'
 
+            self.deactivatepost_tmpl = """!(@(_CONDA_EXE) shell.xonsh deactivatepost)"""
```
Looking at https://github.com/conda/conda/blob/4.5.0rc0/conda/shell/conda.xsh#L24-L38, it seems we need some more code for `xonsh` to properly source the output.

mbargull commented:
https://github.com/conda/conda/blob/4.5.0rc0/tests/test_activate.py will need to be adjusted accordingly once the code is finalized.

njalerikson commented:
The fixes thus far has addressed the nix shells, I haven't looked further at the win shells yet.

Yes, the activate tests need to be updated as well.

I've also addressed two other CSH issues (#6919, #6877) since I was already touching the CSH script. The fix to #6919 is courtesy of @jonathaneichelberger.

mbargull commented on conda/activate.py:
```diff
@@ -23,21 +24,31 @@ class Activator(object):
+    #   deactivatepost      (bool to include deactivate post command)
+    #   reactivatepost      (bool to include reactivate post command)
     #
-    # The value of the CONDA_PROMPT_MODIFIER environment variable holds conda's contribution
```
The prevalent (and linter-checked, https://github.com/conda/conda/blob/4.5.0rc0/setup.cfg#L34) style in the code base uses a 99 character line limit. Please don't enforce foreign rules and revert all those changes.

njalerikson replied:
Didn't realize conda wasn't using the 78 char convention.

mbargull commented on conda/activate.py:
```diff
@@ -71,10 +87,16 @@ def __init__(self, shell, arguments=None):
 
             self.unset_var_tmpl = 'unset %s'
-            self.export_var_tmpl = 'setenv %s "%s"'
+            self.unexport_var_tmpl = 'unsetenv %s'
```
I see no reason to introduce `unexport_var_tmpl`. With your changes you effectively just replaced `unset_var_tmpl` with that. The `unset_var_tmpl` is only used to unset environment variables. If you think `unsetenv` should be used for `csh`, then just you that in `unset_var_tmpl`.

njalerikson replied:
While I agree with this I chose to add it as it makes the solution more distinct in what it's actually unsetting - but alright - I'll revert.

mbargull commented on conda/activate.py:
```diff
@@ -58,10 +69,15 @@ def __init__(self, shell, arguments=None):
+            self.export_var_tmpl = "\\export %s='%s'"
             self.run_script_tmpl = '\\. "%s"'
 
+            self.post_tmpl = dedent("""\
```
Don't overly complicate things, just use
```python
self.post_tmpl = '''eval "$(PS1="${PS1}" "${_CONDA_EXE}" shell.posix %spost)"'''
```

mbargull commented on conda/activate.py:
```diff
@@ -84,9 +106,18 @@ def __init__(self, shell, arguments=None):
 
+            self.post_tmpl = dedent("""\
+                post_pipeline = !(@(_CONDA_EXE) shell.xonsh %spost)
+                post_stdout = _raise_pipeline_error(post_pipeline)
```
`_raise_pipeline_error` is a function defined in https://github.com/conda/conda/blob/4.5.0rc0/conda/shell/conda.xsh#L24-L31. We shouldn't assume it's always available.

njalerikson replied:
Well I would also expect any changes to the individual shell scripts to also make mirrored updates to the applicable `post_tmpl`. This is why I consider this implementation to be relatively dumb.

mbargull commented on conda/activate.py:
```diff
@@ -84,9 +106,18 @@ def __init__(self, shell, arguments=None):
+            self.set_var_tmpl = self.export_var_tmpl
             self.run_script_tmpl = 'source "%s"'
 
+            self.post_tmpl = dedent("""\
```
I'm not big fan of the line-continuation characters in Python code.
I'd rather just
```python
self.post_tmpl = dedent("""
...
""".strip("\n"))
```
or
```python
self.post_tmpl = "\n".join((
    "line1",
    "line2",
))
```

njalerikson replied:
I totally agree, I still haven't quite decided what I prefer. Will reconsider this.

mbargull commented on conda/activate.py:
```diff
@@ -96,9 +127,15 @@ def __init__(self, shell, arguments=None):
             self.run_script_tmpl = '@CALL "%s"'
 
+            self.post_tmpl = dedent("""\
+                @CALL %_CONDA_EXE% shell.cmd.exe %spost\
```
I know this part is WIP, so just a heads-up: use `%%_CONDA_EXE%%`.

mbargull commented on conda/activate.py:
```diff
@@ -234,15 +308,18 @@ def build_activate(self, env_name_or_prefix):
-        activate_scripts = self._get_activate_scripts(prefix)
+        assert 0 <= old_conda_shlvl <= max_shlvl
+
+        args = {}
```
I'd rather not have these unrelated `args = {}` changes in this PR.
It maybe would make sense to do that as part of https://github.com/conda/conda/issues/6595.

njalerikson replied:
It was introduced in part because of the unexport idea but also because I know that the prompt support is currently limited to posix and csh. This change was done for future needs in the update prompt function. But since that's being reverted I guess this will be reverted as well.

mbargull commented on conda/shell/etc/profile.d/conda.csh:
```diff
@@ -4,43 +4,49 @@
-# This block should only be for dev work. Under normal installs, _CONDA_EXE will be templated
-# in at the top of this file.
+# This block should only be for dev work. Under normal installs, _CONDA_EXE
+# will be templated in at the top of this file.
```
Like with the Python code: Just retain the 99 character lines.

mbargull commented on conda/shell/etc/profile.d/conda.csh:
```diff
@@ -4,43 +4,49 @@
             breaksw
         default:
-            $_CONDA_EXE $argv[1-]
+            ${_CONDA_EXE:q} ${argv[1-]:q}
```
I don't know `csh`, so can't really comment on those changes. I guess it makes sense to put that in a separate PR, so that it's easier for a C-shell-knowing person to review.

njalerikson replied:
CSH's `:q` variable modifier properly preserves quoting within the variables. I was trying to do this same trick for the `'${2}' ${argv[3-]}` but due to the sub-shelling it doesn't work (CSH is fun...).

As for another PR to improve review, from my experience with Conda (and just looking at how long CSH issues remain unaddressed in Conda) it will not be easy to find another CSH user.

mbargull commented on conda/shell/etc/profile.d/conda.sh:
```diff
@@ -9,24 +9,25 @@ _conda_set_vars() {
         _CONDA_SHELL_FLAVOR=posh
     else
-        # default to dash; if we run into a problem here, please raise an issue
+        # default to dash; if we run into a problem here, please raise an
```
ditto, character limit

kalefranz commented:
I know I've been absent the last week or so.  Been working on some other projects.  I'm generally onboard with this approach now, even though I may have hinted that I wasn't in #3915.

I do prefer that we try to keep PRs as single-purpose as possible.  And I don't mind breaking PRs into multiple, that must be reviewed and merged in a certain order, if one builds on another.

kalefranz commented:
I'm guessing there are several problems with csh/tcsh support.  I'm glad we're finally looking at it again!

kalefranz commented:
Want to rebase this one again master now and see where we're at?

njalerikson commented:
Woohoo! A successful rebase!

njalerikson commented:
If still desired I'll roll back the incremental minor changes I have introduced and bundle them into a separate PR.

njalerikson commented:
@kalefranz do we want this to be implemented before or after #6518?

njalerikson commented:
I realized that actually all three functions (`activate`, `deactivate`, and `reactivate`) all had cases where a `post` function was necessary.

So I back tracked a little and instead of splitting each function into two phases (e.g. `deactivate` and `deactivatepost`) I introduced a new function called `post`. The `post` function relies on a new environment variable `$CONDA_POST` which is a `old_prefix:new_prefix` syntax. This then allows the `post` command to only deal with fixing the `$PATH` after the necessary `deactivate.d` scripts have been called (as well as unsetting `$CONDA_POST`).

Yes I am still missing the `post_tmpl` for `cmd.exe` and `powershell`. And yes I am still missing tests.

njalerikson commented:
I think I have addressed the issues for Nix shells.

@kalefranz Is this still a PR that is of interest? If so then I'll start hacking on the Powershell and Cmd.exe solutions that are still missing.

njalerikson commented:
Just realized that Powershell isn't actually supported yet. So I guess I'll keep ignoring that one for now.

beenje commented:
Anything I could do to help this PR move forward?

beenje commented:
I'd really like to be able to modify the PATH in activate/deactivate script.
@mbargull is it worth me trying to rebase this PR on master and keep only what is needed?

beenje commented:
Thanks for the reply @msarahan. I'll give it a try!

beenje commented:
Looking at this PR, I wonder if it wouldn't be easier to save the original PATH in an environment variable during the activate and always restore it during deactivate.

That wouldn't allow to modify the PATH in the deactivate.d scripts. But I think what we really want is for the PATH to always be the same as before the activate after deactivate (even when the PATH was modified during the activate.d scripts).

This idea didn't come up when discussing #3915. Is that a bad idea?

beenje commented:
This PR is quite old and the PATH issue has been resolved by https://github.com/conda/conda/pull/8464

My guess is that it should be closed.

msarahan commented:
@njalerikson I'll close this, but I'm happy to reopen if it still needs to be considered.
